File: src/main/java/shane/leetcode/problems/medium/Q791CustomSortString.java
package shane.leetcode.problems.medium;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Runtime 0 ms Beats 100.00% of users with Java
 */
public class Q791CustomSortString {

    @Test
    public void test() {
        assertThat(customSortString("cba", "abcd")).isEqualTo("cbad");
        assertThat(customSortString("bcafg", "abcd")).isEqualTo("bcad");
    }

    public String customSortString(String order, String s) {
        StringBuilder sb = new StringBuilder();
        int[] counts = new int[26];
        for (char c : s.toCharArray()) {
            counts[c - 'a']++;
        }
        for (char c : order.toCharArray()) {
            int count = counts[c - 'a'];
            for (int i = 0; i < count; i++) {
                sb.append(c);
            }
            counts[c - 'a'] = 0;
        }
        for (int i = 0; i < 26; i++) {
            int count = counts[i];
            char c = (char) ('a' + i);
            for (int j = 0; j < count; j++) {
                sb.append(c);
            }
        }
        return sb.toString();
    }

}

File: src/main/kotlin/shane/leetcode/problems/medium/Q791CustomSortString.kt
package shane.leetcode.problems.medium

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.util.function.Consumer

/**
 * Runtime: 258 ms, faster than 56.41% of Kotlin online submissions for Custom Sort String.
 * Memory Usage: 34.5 MB, less than 76.92% of Kotlin online submissions for Custom Sort String.
 */
class Q791CustomSortString {

    @Test
    fun test() {
        assertThat(customSortString("cba", "abcd")).isEqualTo("cbad")
        assertThat(customSortString("cbafg", "abcd")).isEqualTo("cbad")
    }

    fun customSortString(order: String, s: String): String {
        val map: HashMap<Char, Int> = HashMap()
        for (c in s) {
            map.merge(c, 1, Integer::sum)
        }
        val sb: StringBuilder = StringBuilder()
        for (c in order) {
            for (i in 1..map.getOrDefault(c, 0)) {
                sb.append(c)
            }
            map.remove(c)
        }
        map.entries.forEach(Consumer { e ->
            run {
                for (i in 1..map.getOrDefault(e.key, 0)) {
                    sb.append(e.key)
                }
            }
        })
        return sb.toString()
    }
}
